README.md
# tap-trustpilot

This is a [Singer](https://singer.io) tap that produces JSON-formatted data
following the [Singer
spec](https://github.com/singer-io/getting-started/blob/master/SPEC.md).

This tap:

- Pulls raw data from [TrustPilot](https://developers.trustpilot.com/)
- Extracts the following resources:
  - [Business Unit Reviews](https://developers.trustpilot.com/business-units-api#get-a-business-unit's-reviews)
  - [Consumers](https://developers.trustpilot.com/consumer-api#get-the-profile-of-the-consumer(with-#reviews-and-weblinks))
- Outputs the schema for each resource

This tap does _not_ support incremental replication!

## Install

Clone this repository, and then install using setup.py. We recommend using a virtualenv:

```bash
    > python3 -m venv VENV_TAP_TRUSTPILOT
    > source venv/bin/activate
    > python setup.py install
    OR
    > cd .../tap-trustpilot
    > pip install .
```

### Configuration

Create a `config.json` file that looks like this:

```
{
    "api_key": "...",
    "business_unit_ids": "123abc, 345fgh, tyuie456"
}
```

## Discovery Mode

Run the Tap in Discovery Mode. This creates a catalog.json for selecting objects/fields to integrate:
    
```bash
> tap-trustpilot --config config.json --discover > catalog.json
```
   
See the Singer docs on discovery mode [here](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#discovery-mode).

## Sync Mode

Run the Tap in Sync Mode (with catalog) and [write out to state file](https://github.com/singer-io/getting-started/blob/master/docs/RUNNING_AND_DEVELOPING.md#running-a-singer-tap-with-a-singer-target)

For Sync mode:
```bash
> tap-trustpilot --config config.json --catalog catalog.json > state.json
```

---

Copyright &copy; 2022 Stitch


CHANGELOG.md
# Changelog

## 0.2.0
   * Fix Discovery, Sync and bulk endpoint [#1] (https://github.com/singer-io/tap-trustpilot/pull/1)
   * Extract latest data using orderBy desc [#2] (https://github.com/singer-io/tap-trustpilot/pull/2)
   * Add README.md and code comments [#3] (https://github.com/singer-io/tap-trustpilot/pull/3)
   * Support for multiple Business Unit IDs in config [#4] (https://github.com/singer-io/tap-trustpilot/pull/4)

tap_trustpilot/client.py
import requests
from singer import metrics
import singer
import backoff

LOGGER = singer.get_logger()

BASE_URL = "https://api.trustpilot.com/v1"
AUTH_URL = "{}/oauth/oauth-business-users-for-applications/accesstoken".format(BASE_URL)


class RateLimitException(Exception):
    pass


class UnauthorisedException(Exception):
    pass


def _join(a, b):
    return a.rstrip("/") + "/" + b.lstrip("/")


class Client(object):
    def __init__(self, config):
        self.user_agent = config.get("user_agent")
        self.session = requests.Session()

        self.api_key = config['api_key']
        self._token = None

    def validate_api_key(self):
        """
            Function validate_api_key confirms whether the api_key provided in config.json is valid or not
            hits /business-units/all endpoint to validate api_key
            business_unit_id and tap_stream_id are being sent as empty strings since these are not required
        """

        self.GET({'path': '/business-units/all', 'business_unit_id': ''}, '')

    def prepare_and_send(self, request):
        if self.user_agent:
            request.headers["User-Agent"] = self.user_agent

        # request.headers['Authorization'] = 'Bearer {}'.format(self._token)
        request.headers['apikey'] = self.api_key
        request.headers['Content-Type'] = 'application/json'

        return self.session.send(request.prepare())

    def url(self, path, business_unit_id=''):
        # parameterizing business unit ID to support multiple business unit IDs TDL-19427
        joined = _join(BASE_URL, path)
        return joined.replace(':business_unit_id', business_unit_id)

    def create_get_request(self, path, business_unit_id, **kwargs):
        return requests.Request(method="GET", url=self.url(path, business_unit_id), **kwargs)

    def create_post_request(self, path, payload, **kwargs):
        return requests.Request(method="POST", url=self.url(path), data=payload, **kwargs)

    @backoff.on_exception(backoff.expo, RateLimitException, max_tries=10, factor=2)
    @backoff.on_exception(backoff.expo, requests.Timeout, max_tries=10, factor=2)
    def request_with_handling(self, request, tap_stream_id):
        with metrics.http_request_timer(tap_stream_id) as timer:
            response = self.prepare_and_send(request)
            timer.tags[metrics.Tag.http_status_code] = response.status_code
        if response.status_code in [429, 503]:
            raise RateLimitException()
        return self.check_for_http_error(response)

    @staticmethod
    def check_for_http_error(resp_object):
        """
        TrustPilot API has 100k record limit for reviews stream
        below condition should handle Pagination limit exceeded error if page value is more than 1000
        sends an empty response
        """
        if resp_object.status_code == 400 and resp_object.json().get('details') == "Pagination limit exceeded.":
            LOGGER.warning("400 Bad Request, Pagination limit exceeded.")
            return []
        """
        send empty response when given business_unit id is invalid or malformed
        """
        if resp_object.status_code == 400 and (resp_object.json().get('message') ==
                                               "The given business unit id was malformed" or
                                               resp_object.json().get('details') == 'Error: valid unitId is required'):
            return []
        """
        following condition handles unauthorised error..validates apiKey and throws exception if it is not valid
        """
        if resp_object.status_code == 401:
            raise UnauthorisedException("Unauthorised...Invalid ApiKey")
        resp_object.raise_for_status()
        return resp_object.json()

    def GET(self, request_kwargs, *args, **kwargs):
        req = self.create_get_request(**request_kwargs)
        return self.request_with_handling(req, *args, **kwargs)

    def POST(self, request_kwargs, *args, **kwargs):
        req = self.create_post_request(**request_kwargs)
        return self.request_with_handling(req, *args, **kwargs)


tap_trustpilot/schemas.py
#!/usr/bin/env python3
import os
import json
import singer
from singer import utils,metadata
from tap_trustpilot.streams import PK_FIELDS, all_streams


def get_abs_path(path):
    return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)


def load_schema(tap_stream_id):
    path = "schemas/{}.json".format(tap_stream_id)
    return utils.load_json(get_abs_path(path))


def load_and_write_schema(tap_stream_id):
    schema = load_schema(tap_stream_id)
    singer.write_schema(tap_stream_id, schema, PK_FIELDS[tap_stream_id])

def get_schemas():
    """ Load schemas from schemas folder """
    schemas = {}
    field_metadata = {}

    # Iterate over list pf objects of different streams
    for stream_metadata in all_streams:
        path = get_abs_path(f'schemas/{stream_metadata.stream_name}.json')
        with open(path, encoding='utf-8') as file:
            schema = json.load(file)
        schemas[stream_metadata.stream_name] = schema

        mdata = metadata.get_standard_metadata(
            schema=schema,
            key_properties=stream_metadata.key_properties,
            replication_method=stream_metadata.replication_method,
            valid_replication_keys=stream_metadata.replication_keys
        )
        field_metadata[stream_metadata.stream_name] = mdata

    return schemas, field_metadata


tap_trustpilot/__init__.py
#!/usr/bin/env python3
import singer
from singer import utils
from singer.catalog import Catalog
from tap_trustpilot import streams as streams_
from tap_trustpilot.context import Context
from tap_trustpilot import schemas
from tap_trustpilot.discover import discover

REQUIRED_CONFIG_KEYS = [
    "api_key",
    "business_unit_ids"
]

LOGGER = singer.get_logger()

def output_schema(stream):
    schema = schemas.load_schema(stream.tap_stream_id)
    pk_fields = schemas.PK_FIELDS[stream.tap_stream_id]
    singer.write_schema(stream.tap_stream_id, schema, pk_fields)


def sync(ctx):
    bu_ids = ctx.config.get('business_unit_ids').replace(" ", "").split(",")
    # iterating through each business unit id to extract data
    for bu_id in bu_ids:
        LOGGER.info(f"Extracting data for business_unit_id {bu_id}")
        ctx.config['business_unit_id'] = bu_id

        # fetch state value for last sync if value is None do
        # extraction for all the streams
        currently_syncing = ctx.state.get("currently_syncing")
        start_idx = streams_.all_stream_ids.index(currently_syncing) \
            if currently_syncing else 0
        stream_ids_to_sync = [cs.tap_stream_id for cs in ctx.catalog.streams
                              if cs.is_selected()]
        streams = [s for s in streams_.all_streams[start_idx:]
                   if s.tap_stream_id in stream_ids_to_sync]

        for stream in streams:
            # store state value for current stream
            ctx.state["currently_syncing"] = stream.tap_stream_id
            output_schema(stream)
            ctx.write_state()
            stream.sync(ctx)
        # make the state value as None after all streams have passed
        ctx.state["currently_syncing"] = None
        ctx.write_state()

@utils.handle_top_exception(LOGGER)
def main():
    args = utils.parse_args(REQUIRED_CONFIG_KEYS)
    ctx = Context(args.config, args.state)
    if args.discover:
        discover(ctx).dump()
        print()
    else:
        ctx.catalog = Catalog.from_dict(args.catalog.to_dict()) \
            if args.catalog else discover(ctx)
        sync(ctx)

if __name__ == "__main__":
    main()


tap_trustpilot/discover.py
from singer.catalog import Catalog, CatalogEntry, Schema
from tap_trustpilot import schemas

def check_credentials_are_authorized(ctx):
    """
    validates access_key
    """
    ctx.client.validate_api_key()

def discover(ctx):
    """
    Constructs a singer Catalog object based on the schemas and metadata.
    """
    check_credentials_are_authorized(ctx)
    discover_schemas, field_metadata = schemas.get_schemas()
    streams = []
    for stream_name, raw_schema in discover_schemas.items():
        schema = Schema.from_dict(raw_schema)
        mdata = field_metadata[stream_name]
        streams.append(
            CatalogEntry(
                tap_stream_id=stream_name,
                stream=stream_name,
                schema=schema,
                key_properties=schemas.PK_FIELDS[stream_name],
                metadata=mdata
            )
        )
    return Catalog(streams)

tap_trustpilot/streams.py
import json

import singer
from tap_trustpilot import transform

LOGGER = singer.get_logger()
PAGE_SIZE = 100
CONSUMER_CHUNK_SIZE = 1000

class IDS(object):
    BUSINESS_UNITS = "business_units"
    REVIEWS = "reviews"
    CONSUMERS = "consumers"

stream_ids = [getattr(IDS, x) for x in dir(IDS)
              if not x.startswith("__")]

PK_FIELDS = {
    IDS.BUSINESS_UNITS: ["id"],
    IDS.REVIEWS: ["business_unit_id", "id"],
    IDS.CONSUMERS: ["id"],
}

class Stream(object):
    def __init__(self, tap_stream_id, path,
                 returns_collection=True,
                 collection_key=None,
                 custom_formatter=None):
        self.tap_stream_id = tap_stream_id
        self.path = path
        self.returns_collection = returns_collection
        self.collection_key = collection_key
        self.custom_formatter = custom_formatter or (lambda x: x)

    def metrics(self, records):
        with singer.metrics.record_counter(self.tap_stream_id) as counter:
            counter.increment(len(records))

    def write_records(self, records):
        singer.write_records(self.tap_stream_id, records)
        self.metrics(records)

    def format_response(self, response):
        if self.returns_collection:
            if self.collection_key:
                records = (response or {}).get(self.collection_key, [])
            else:
                records = response or []
        else:
            records = [] if not response else [response]

        return self.custom_formatter(records)

    def transform(self, ctx, records):
        schema = ctx.catalog.get_stream(self.tap_stream_id).schema.to_dict()
        transformed = [transform.transform(item, schema) for item in records]
        return transformed


class BusinessUnits(Stream):

    # tap_stream_id = "business_units"
    stream_name = "business_units"
    key_properties = ['id']
    replication_keys  = None
    replication_method = "FULL_TABLE"
    params = {}

    def raw_fetch(self, ctx, business_unit_id):
        # parameterizing business unit ID to support multiple business unit IDs TDL-19427
        return ctx.client.GET({"path": self.path, "business_unit_id": business_unit_id}, self.tap_stream_id)

    def sync(self, ctx):
        """
        Extracts business unit profile for a given business_unit_id
        Logs a warning message if business_unit_id is invalid
        """
        business_unit_id = ctx.config['business_unit_id']

        resp = self.raw_fetch(ctx, business_unit_id)
        if resp:
            resp['id'] = business_unit_id

            business_units = self.transform(ctx, [resp])

            ctx.cache["business_unit"] = business_units[0]

            self.write_records([ctx.cache["business_unit"]])
        else:
            LOGGER.warning("Business Unit {} was not found!".format(business_unit_id))


class Paginated(Stream):
    @staticmethod
    def get_params(page):
        return {
            "page": page,
            "perPage": PAGE_SIZE,
            "orderBy": "createdat.desc"
        }

    def _sync(self, ctx):
        """
        Extracts data in paginated streams
        default page size is 100 and cannot exceed more than 100 due to trust pilot API restrictions
        """
        business_unit_id = ctx.config['business_unit_id']

        page = 1
        while True:
            LOGGER.info("Fetching page {} for {}".format(page, self.tap_stream_id))
            params = self.get_params(page)
            resp = ctx.client.GET({"path": self.path, "params": params, "business_unit_id":business_unit_id}, self.tap_stream_id)
            raw_records = self.format_response(resp)

            for raw_record in raw_records:
                raw_record['business_unit_id'] = business_unit_id

            records = self.transform(ctx, raw_records)
            self.write_records(records)
            yield records

            if len(records) < PAGE_SIZE:
                break

            page += 1


class Reviews(Paginated):

    key_properties = ["business_unit_id", "id"]
    stream_name = "reviews"
    replication_keys = None
    replication_method = "FULL_TABLE"
    params = {}

    @staticmethod
    def add_consumers_to_cache(ctx, batch):
        for record in batch:
            consumer_id = record.get('consumer', {}).get('id')
            if consumer_id is not None:
                ctx.cache['consumer_ids'].add(consumer_id)

    def sync(self, ctx):
        """
        Extracts data for consumer reviews in paginated manner
        Caches consumer id for each review
        """
        ctx.cache['consumer_ids'] = set()
        for batch in self._sync(ctx):
            self.add_consumers_to_cache(ctx, batch)


class Consumers(Stream):

    key_properties = ['id']
    stream_name = "consumers"
    replication_keys = None
    replication_method = "FULL_TABLE"
    params = {}

    def sync(self, ctx):
        """
        extracts consumer profiles using bulk endpoint.
        Chunks the Consumer IDs set to smaller list of size 1000
        """
        business_unit_id = ctx.config['business_unit_id']

        total = len(ctx.cache.get('consumer_ids',[]))
        LOGGER.info("Total Consumer profiles to be extracted {}".format(total))

        # chunk list of consumer IDs to smaller lists of size 1000
        consumer_ids = list(ctx.cache.get('consumer_ids',[]))
        chunked_consumer_ids = [consumer_ids[i: i+CONSUMER_CHUNK_SIZE] for i in range(0, len(consumer_ids),
                                                                                      CONSUMER_CHUNK_SIZE)]

        for idx, consumer_id_list in enumerate(chunked_consumer_ids):
            LOGGER.info("Fetching consumer profiles page {} of {}".format(idx + 1, len(chunked_consumer_ids)))
            resp = ctx.client.POST({"path": self.path, "payload": json.dumps({"consumerIds": consumer_id_list})},
                                   self.tap_stream_id)

            raw_records = self.format_response([resp])
            raw_records = list(raw_records[0].get('consumers', {}).values())
            for raw_record in raw_records:
                raw_record['business_unit_id'] = business_unit_id

            records = self.transform(ctx, raw_records)
            self.write_records(records)


business_units = BusinessUnits(IDS.BUSINESS_UNITS, "/business-units/:business_unit_id/profileinfo")
all_streams = [
    business_units,
    Reviews(IDS.REVIEWS, '/business-units/:business_unit_id/reviews', collection_key='reviews'),
    Consumers(IDS.CONSUMERS, '/consumers/profile/bulk')
]
all_stream_ids = [s.tap_stream_id for s in all_streams] 
